sympy/codegen/ast.py

```diff
@@ -399,3 +400,405 @@ def body(self):
+
+        Examples
+        --------
+        >>> from sympy.codegen.ast import Type
+        >>> Type('integer').cast_check(3.0) == 3
+        True
+        >>> Type('float32').cast_check(1e-40)  # doctest: +ELLIPSIS
+        Traceback (most recent call last):
+          ...
+        ValueError: Minimum value for data type bigger than new value.
+
+        """
+        from sympy.functions.elementary.complexes import im, re
+        def lim(type_name, key):
+            return (limits or self.default_limits).get(type_name, {}).get(key)
+        val = _sympify(value)
+
+        name = (precision_targets or self.default_precision_targets).get(self.name, self.name)
+        ten = Integer(10)
+        if rtol is None:
+            exp10 = lim(name, 'decimal_dig')
+            rtol = 1e-15 if exp10 is None else ten**(-exp10)
+
+        if atol is None:
+            if rtol == 0:
+                exp10 = lim(name, 'decimal_dig')
+                atol = 1e-15 if exp10 is None else 10**(-exp10)
+            else:
+                atol = 0
+
+        def tol(num):
+            return atol + rtol*abs(num)
+
+        caster = lambda x: x  # identity
+        _min, _max, _tiny = -oo, oo, 0  # undefined precision
+
+        if name == 'integer':
+            caster = int
+        elif name.startswith('int'):
+            nbits = int(name.split('int')[1])
+            _min, _max = -2**(nbits - 1), 2**(nbits - 1) - 1
+            caster = int
+        elif name.startswith('uint'):
+            nbits = int(name.split('uint')[1])
+            _min, _max = 0, 2**nbits - 1
+            caster = int
+        elif name.startswith('float'):
+            _max = +lim(name, 'max')
+            _min = -lim(name, 'max')
+            _tiny = lim(name, 'tiny')
+            dec_dig = lim(name, 'decimal_dig')
+            val = Float(str(val), dec_dig+3)
+            caster = lambda x: Float(str(x.evalf(dec_dig)), dec_dig+3)
+
+        if name.startswith('complex'):
+            if name == 'complex':
+                nbits = 128  # assume double precision as underlying data
+            else:
+                nbits = int(name.split('complex')[1])
+            corresponding_float = 'float%d' % (nbits // 2)
+            _max = +lim(corresponding_float, 'max')
+            _tiny = lim(corresponding_float, 'tiny')
+            dec_dig = lim(corresponding_float, 'decimal_dig')
+            caster = lambda x: (
+                Float(str(re(x).evalf(dec_dig)), dec_dig+3) +
+                Float(str(im(x).evalf(dec_dig)), dec_dig+3)*1j
+            )
+            if abs(re(val)) > _max or abs(im(val)) > _max:
+                raise ValueError("Maximum value exceeded for data type.")
+            if abs(re(val)) < _tiny or abs(im(val)) < _tiny:
+                raise ValueError("Minimum (absolute) value for data type bigger than new value.")
+            new_val = caster(re(val)) + 1j*caster(im(val))
```

Comment:
I mean it seems like your definition is wrong. Should it just be `new_val = caster(val)`?